### ISSUE TITLE
[Feature Request] Turning OpenOutpaint into an A1111 Extension

### ISSUE THREAD

**Issue by wywywywy** [zero01101/openOutpaint#109]
**Is your feature request related to a problem? Please describe.**
OpenOutpaint is a great idea. But having to run another app does not make the workflow seamless.

**Describe the solution you'd like**
Add OpenOutpaint as an extension to A1111. This will become a new "Outpaint" tab in the A1111 UI.

**Describe alternatives you've considered**
I guess you could edit the `webui-user.bat` file to run both A1111 and OpenOutpaint at the same time, but it is still not really seamless.

**Additional context**
It should be simple (I think) to build an iframe in a tab. So the changes to the codebase shouldn't be too big.

In the future it could have additional features such as sending a generated image from the "txt2img" tab to the "Outpaint" tab.

**Comment by seijihariki** [comment]
Could be a thing I guess? Haven't seen how extensions work but as we only need some basic hosting for some files, an iframe should work just fine.

Though updates may be a bit hard to apply depending on browser caching. Maybe add something to handle that. 

Also should lock the host field and have a button to allow for iframe refreshing, as we don't really have a reset button for some functionalities (canvas expansion).

We should also not pollute the webui's local storage or IndexedDB, so starting some parallel file hosting would be preferable.

**Comment by seijihariki** [comment]
We could probably achieve complete integration sending resources between the two interfaces using messaging, actually.

**Comment by seijihariki** [comment]
@zero01101, what do you think? We can use a separate repo for this, I think, for the actual extension and python code involved. For communication with the webui itself, we can use messaging between webui and openOutpaint: https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage.

**Comment by zero01101** [comment]
hah, i actually tried messing with just a very remedial attempt at getting an iframe in there like suggested; my python is admittedly worse than my javascript, and i couldn't figure out how in blue blazes to get it to embed a git clone of openOutpaint, but just as a stupid test got the github.io-hosted one "working"... not sure where the button graphics went though...  

![image](https://user-images.githubusercontent.com/1649724/208316690-f040d7d2-9abf-412e-ab5d-7a862385dacd.png)
 
 but agreed, i definitely think it'd be its own separate repo; i've got the `install.py` successfully cloning openOutpaint to the extension's directory already but that's, uh, pretty "step one" kinda stuff :/  
 
 i'm _sure_ you'll be shocked to hear that i've never used `postMessage()` lmao

**Comment by seijihariki** [comment]
> hah, earlier this morning i actually tried messing with just a very remedial attempt at getting an iframe in there like suggested; my python is admittedly worse than my javascript, and i couldn't figure out how in blue blazes to get it to embed a git clone of openOutpaint, but just as a stupid test got the github.io-hosted one "working"... not sure where the tool icons went though...
> 
> ![image](https://user-images.githubusercontent.com/1649724/208316690-f040d7d2-9abf-412e-ab5d-7a862385dacd.png)

Hey, that's good! Can you try sending this to a new repo, so I can try doing it from there? And is the github-sites version of openOutpaint broken in terms of icon loading?  That seems a bit strange to me.

> but agreed, i definitely think it'd be its own separate repo; i've got the `install.py` successfully cloning openOutpaint to the extension's directory already but that's, uh, pretty "step one" kinda stuff :/
> 
> i'm _sure_ you'll be shocked to hear that i've never used `postMessage()` lmao

Not really, as it has only some pretty niche applications. Primarily for inter-tab communication and parent-child iframe communication. I used this only once in my life for very basic stuff.

**Comment by zero01101** [comment]
>Hey, that's good!

lmao you will shortly discover that it is _not_ ;) 

invited you as a direct collaborator to https://github.com/zero01101/openOutpaint-webUI-extension :)

**Comment by wywywywy** [comment]
It will probably be more suitable to make use of git submodules in the new git repo instead of trying to get git clone in python to work

**Comment by seijihariki** [comment]
Are submodules automatically downloaded on extension clone? If so, I agree.

**Comment by seijihariki** [comment]
> > And is the github-sites version of openOutpaint broken in terms of icon loading? That seems a bit strange to me.
> 
> sure is :| super bonkers, right?
> 
> invited you as a direct collaborator to https://github.com/zero01101/openOutpaint-webUI-extension :)

Just checked, this seems to be an issue of relative paths. openOutpaint, for now, assumes that it is being run on the server root, so it uses "/res" as a first part for icon loading. The thing is, if we use relative paths, the configuration page icons will break, as it is on another path. We can use "../res" for resources on the config page, but then we duplicate icon loading code, as CSS would not be able to be shared.

We could also move configuration.html from pages to root, but then we start filling the root with html pages... Not sure what is the better approach.

**Comment by seijihariki** [comment]
Added issue for discussion of implementation for the extension on the separate repository:
https://github.com/zero01101/openOutpaint-webUI-extension/issues/1

**Comment by zero01101** [comment]
ok lmao wow that got really good at making me feel like a complete pillock lol

anyway, no more install script, submodules work finally, not "magically" installed unless you `git clone --recursive` or something, so i just added a submodule update routine to the script itself

definitely makes sense re: missing icons (and makes even more sense now that it's running off of the local submodule lol) but yeah, i'm not exactly sure what the best course of action is for that... 

all that being said, it does seem to work as-is right now just by manually installing the extension from the repo url 👍

**Comment by seijihariki** [comment]
Is it already working from local install or does it still use GitHub pages?

**Comment by zero01101** [comment]
local install via git submodule :D

**Comment by seijihariki** [comment]
> local install via git submodule :D

Then will have to use relative paths probably; shouldn't be too big a change to fix the images... Just a general search and replace of /res/ to res/, and custom css for the X on the configuration.html file.

**Comment by zero01101** [comment]
ooooo <3

**Comment by seijihariki** [comment]
Should do the same. If you prefer, ./res also works.

**Comment by zero01101** [comment]
aha, seems not as much as i'd like :| 
![image](https://user-images.githubusercontent.com/1649724/208325694-d3ec885b-3b0f-469e-ab3c-887353ffa799.png)

**Comment by seijihariki** [comment]
Oh, the css are really loading relative files from their directories? That's... Suboptimal. Will probably have to refer to ../../res then.

**Comment by seijihariki** [comment]
In the upside, probably no duplication will actually need to happen for configuration.html to work.

**Comment by zero01101** [comment]
yep, just tried exactly that for files in `ui/tool` (and simply `../res` for files in `ui`) and it seems to have worked :)

**Comment by zero01101** [comment]
🎉 
![image](https://user-images.githubusercontent.com/1649724/208326393-82b5019f-e4f6-4113-b5a4-008134af75dc.png)

**Comment by seijihariki** [comment]
@wywywywy do you know how it works to add an extension to the main webui repo? Do we have to request it somewhere?

**Comment by wywywywy** [comment]
Yea you gotta ping @ClashSAN to get your extension added to this page: https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Extensions-index

**Comment by wywywywy** [comment]
Once done it'll probably be a good idea to go onto to reddit and promote it a bit. To get more eyeballs on it

**Comment by seijihariki** [comment]
> Yea you gotta ping @ClashSAN to get your extension added to this page: https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Extensions-index

Oh, okay!

@wywywywy Give it a look now. It should be mostly working.

**Comment by wywywywy** [comment]
Ok I don't have much time right now, but after a quick test, I can say that it looks good!

For the first release, I think it needs a couple of changes to be complete:

1. It doesn't really need to ask the user for the server URL. It can just take it from the address bar because when acting as an extension, they'll always be the same
1. When doing "Send to openOutpaint", it should also send the prompt

And going forward, it'd be really nice to have the ability to send the canvas to img2img, so that the user can further process it (e.g. with SD Upscale). Maybe "Send to extras" too.

Another thing is when using the "Save" button, it doesn't save the metadata. But it's a general issue, not specific to the extension.

**Comment by seijihariki** [comment]
Yes, I agree with most of the points. The host really should be locked, and send to img2img is on the way. This is most of a first look to check for basic functionalities and compatibility issues.

Save not saving metadata is a complicated topic... Considering what would happen when a downloaded canvas contains multiple generations, or even worse, multiple layers.

Even if we only implement this for image generation saving and/or generated resource downloading, this would need to be implemented as manual editing of the image binary exif data, which would be... Unpleasant to say the least.

**Comment by wywywywy** [comment]
Yea you're right. Saving metadata is not all that important so let's forget about that for the time being

**Comment by seijihariki** [comment]
Just open an issue for this metadata stuff so we don't forget. It probably will stay there for a while, but at least it is on a todo somewhere.

**Comment by zero01101** [comment]
kind of irrelevant but regarding the extensions index, seems @ClashSAN may already be aware?  https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Extensions#test

i'm a particular fan of the last entry and am very looking forward it its release :D

**Comment by seijihariki** [comment]
Oh god, it's almost scary how effective whatever he is doing is in finding extensions! The new repo was barely twelve hours old, and he included it in the wiki.

**Comment by zero01101** [comment]
> Not sure if we will have an update soon, though, considering the last commit date.

MAN MY COWORKER ALWAYS GOES ON ABOUT "OH THIS REPO HASN'T BEEN COMMITTED TO IN 6 YEARS" and i'm just like MAYBE THAT MEANS IT'S PERFECT _ALREADY_


![lil bad pokerface](https://user-images.githubusercontent.com/1649724/208553114-6e694313-dab6-4b35-8d72-04c9fb8bdf8c.png)

**Comment by seijihariki** [comment]
>     1. It doesn't really need to ask the user for the server URL. It can just take it from the address bar because when acting as an extension, they'll always be the same
> 
>     2. When doing "Send to openOutpaint", it should also send the prompt
> 

These two are done. Need to see what we can do about model selection on openOutpaint (how we will handle changes in model on webui and openoutpaint). Sending to image is on the way. Will probably use the select tool for that functionality. (also need to fix some bugs on select tool with the layer system)

**Comment by seijihariki** [comment]
For the model stuff, I'm thinking of disabling model changing from the openoutpaint ui.

**Comment by zero01101** [comment]
definitely agree that just disabling the model selector in the openOutpaint extension is probably the safest and most "obvious" way to do that

**Comment by seijihariki** [comment]
I think, in general, it's safe to consider the current state of the extension as fully functional and stable enough. We can ask @ClashSAN if there are any problems with the way we are updating our submodules and to add it to the main list. Also, create a simple README for the extension repo.

**Comment by zero01101** [comment]
i added a tiny little readme with a quick screenshot; it's pretty bare however :/

**Comment by wywywywy** [comment]
It'd be a good idea to raise a feature request issue on the A1111 repo to get some visibility & tracking

**Comment by seijihariki** [comment]
@wywywywy Seems like (at)ClashSAN (will not tag him to avoid sending too many notifications) has taken a look at our repo: https://github.com/zero01101/openOutpaint-webUI-extension/issues/1#issuecomment-1363358065

Will keep this issue open for a bit, even if/after the extension is officially added, and close it after implementing proper model selection handling and send to img2img.

**Comment by zero01101** [comment]
WE IN THE BIG LEAGUES NOW 🎉 

![image](https://user-images.githubusercontent.com/1649724/209248011-637b39e9-c0d5-44b7-984b-abf9a0cd057a.png)

![image](https://user-images.githubusercontent.com/1649724/209248083-ec80bff4-8612-4daa-8e38-530adcd148b7.png)

**Comment by seijihariki** [comment]
> WE IN THE BIG LEAGUES NOW 🎉 
> 
> ![image](https://user-images.githubusercontent.com/1649724/209248011-637b39e9-c0d5-44b7-984b-abf9a0cd057a.png)
> 
> ![image](https://user-images.githubusercontent.com/1649724/209248083-ec80bff4-8612-4daa-8e38-530adcd148b7.png)

🎉🎉🎉

**Comment by zero01101** [comment]
think i can probably safely close this issue now ;)
